packages/text-decorator/src/text-decorator.ts

```diff
@@ -1,6 +1,248 @@
+  replace: string | ReactElement;
+}
+
+function hasUnbalancedBrackets(str: string) {
+  let bracketCount = 0;
+  for (let i = 0; i < str.length; ++i) {
+    const c = str.charAt(i);
+    if (c === '[') {
+      ++bracketCount;
+    }
+    else if (c === ']') {
+      if (bracketCount <= 0) return true;
+      --bracketCount;
+    }
+  }
+  return bracketCount !== 0;
+}
+
+// based on https://stackoverflow.com/a/6969486
+function escapeRegExp(str: string) {
+  const regex = hasUnbalancedBrackets(str)
+                    // allow '.', '?', but escape brackets
+                  ? /[\-\[\]\/\{\}\(\)\*\+\\\^\$\|]/g
+                    // allow '.', '?', '[', ']'
+                  : /[\-\/\{\}\(\)\*\+\\\^\$\|]/g;
+  return str.replace(regex, "\\$&");
+}
+
+function generateRegEx(words: string[]) {
+  const escaped = words.map(expr => escapeRegExp(expr))
+                      .filter(expr => !!expr);
+  return new RegExp(`(?:^|\\b)(${escaped.join('|')})(?=\\b|$)`, 'gi');
+}
+
+export function decorateHtml(input: string, options: IDecorateHtmlOptions) {
+  const textRuns: ITextRun[] = [];
+  // regex for matching glossary words
+  const regex = generateRegEx(options.words);
+  const parseOptions: ParserOptions = { sourceCodeLocationInfo: true };
+  const fragment: Node = parse5.parseFragment(input, parseOptions) as Node;
+
+  const identifyTextRuns = (node: Node) => {
+    const nodeAsParent = node as any as ParentNode;
+    if (node.nodeName === '#text') {
+      const textNode = node as any as TextNode;
+      const location = textNode.sourceCodeLocation as Location;
+      textRuns.push({ start: location.startOffset, end: location.endOffset });
+    }
+    else if (nodeAsParent.childNodes && nodeAsParent.childNodes.length) {
+      nodeAsParent.childNodes.forEach((node) => identifyTextRuns(node));
+    }
+    return node;
+  };
+  identifyTextRuns(fragment);
+
+  // replace text in text runs where appropriate
+  let result = input;
+  for (let i = textRuns.length - 1; i >= 0; --i) {
+    const textRun = textRuns[i];
+    const text = result.substring(textRun.start, textRun.end);
+    const newText = text.replace(regex, options.replace);
+    if (newText !== text) {
+      const prefix = textRun.start > 0 ? result.substring(0, textRun.start) : '';
+      const suffix = textRun.end < result.length ? result.substring(textRun.end) : '';
+      result = `${prefix}${newText}${suffix}`;
+    }
+  }
+  return result;
+}
+
+export function decorateReact(input: ReactNode, options: IDecorateReactOptions): ReactNode {
+  // regex for matching glossary words
+  const regex = generateRegEx(options.words);
+  // generates replacement React element for matched words
+  const replaceElement = (match: string, replace: ReactElement, index: number = 0) => {
+    const children = replace.props.children;
+    const newChildren = (children == null) || (children.length === 0)
+                          ? match
```

Comment:
If I was searching for "foo" in a react element with no children, wouldn't this insert "foo" as a child element?

I will keep reading, but this is my current interpretation.

Reply:
The goal of this code (and it's also the answer to another of your questions below) is to allow the client to leave off the `$1` and still have the right thing happen. Since the only use case we currently envision involves decorating existing text, one can argue that the `$1` is implied, and so if the replacement JSX is for a single element with no children then it is (according to this theory) reasonable to assume that the new element is meant to wrap the original matched string. In this case, using the `$1` would only be necessary when the replacement text is something other than the original match (like if it were also wrapping it in quotes). That said, I could also see an argument that this is a too-clever-by-half feature that could potentially preclude some valid use case we haven't thought of yet, so I'm open to discarding it if we collectively decide that the convenience of being able to drop the `$1` isn't worth it.